src/models/util/ParseCalChart31.ts

```diff
@@ -0,0 +1,355 @@
+ * POINT_LABEL_FLIP = INGL_LABL , BigEndianInt32(Sizeof(POINT_LABEL_FLIP_DATA)),
+ *                    POINT_LABEL_FLIP_DATA ;
+ * POINT_LABEL_FLIP_DATA = { BigEndianInt8( label flipped ) } ;
+ * CONTINUITY = INGL_CONT , BigEndianInt32(Sizeof(CONTINUITY_DATA)) ,
+ *              CONTINUITY_DATA ;
+ * CONTINUITY_DATA = CONTINUITY_INDEX , CONTINUITY_NAME , CONTINUITY_TEXT ;
+ * CONTINUITY_INDEX = BigEndianInt8( index );
+ * CONTINUITY_NAME = Null-terminated char* ;
+ * CONTINUITY_TEXT = Null-terminated char* ;
+ * SHEET_END = INGL_END , INGL_SHET ;
+ * SHOW_END = INGL_END , INGL_SHOW ;
+ * INGL_INGL = 'I','N','G','L' ;
+ * INGL_GURK = 'G','U','R','K' ;
+ * INGL_SHOW = 'S','H','O','W' ;
+ * INGL_SHET = 'S','H','E','T' ;
+ * INGL_SIZE = 'S','I','Z','E' ;
+ * INGL_LABL = 'L','A','B','L' ;
+ * INGL_MODE = 'M','O','D','E' ;
+ * INGL_DESC = 'D','E','S','C' ;
+ * INGL_NAME = 'N','A','M','E' ;
+ * INGL_DURA = 'D','U','R','A' ;
+ * INGL_POS  = 'P','O','S',' ' ;
+ * INGL_SYMB = 'S','Y','M','B' ;
+ * INGL_TYPE = 'T','Y','P','E' ;
+ * INGL_REFP = 'R','E','F','P' ;
+ * INGL_CONT = 'C','O','N','T' ;
+ * INGL_PCNT = 'P','C','N','T' ;
+ * INGL_END  = 'E','N','D',' ' ;
+ * Some assumptions:
+ * strings are ascii 8-bit chars.
+ * Error if LABEL_DATA does not contain N null terminated strings, where N ==
+ * number of marchers
+ * Error if POSITION_DATA does not contain N*2 values, where N == number of
+ * marchers
+ * Error if REF_POSITION_DATA does not contain N*2 + 1 values, where N == number
+ * of marchers
+ * Error if POINT_SYMBOL_DATA does not contain N values, where N == number of
+ * marchers
+ * Error if POINT_CONT_INDEX_DATA does not contain N values, where N == number
+ * of marchers
+ * Error if POINT_LABEL_FLIP_DATA does not contain N values, where N == number
+ * of marchers
+ * If REF_POSITION is not supplied, all reference points are assumed to be set
+ * to the point value
+ * If POINT_SYMBOL is not supplied, all points assumed to be symbol 0
+ * If POINT_CONT_INDEX is not supplied, all points assumed to be index 0
+ * If POINT_LABEL_FLIP is not supplied, all points assumed to be not flipped
+ */
+
+import Show from '../Show';
+import StuntSheet from '../StuntSheet';
+import StuntSheetDot from '../StuntSheetDot';
+import { readInt32,
+  readInt16,
+  readStringTillEnd,
+  readArrayOfStringsTillEnd,
+  readFourCharCode,
+  calChart3To4ConvertX,
+  calChart3To4ConvertY } from './ParseCalChart3Utils';
+import { ParseCalChart } from './ParseCalChart';
+
+export class ParseCalChart31 implements ParseCalChart {
+  private numberDots: number = 0;
+
+  ParseShow(inputBuffer: ArrayBuffer): Show {
+    // we know the header for a CalChart3.1 show is 8 bytes.  Remove it.
+    const buffer = new DataView(inputBuffer, 0, inputBuffer.byteLength);
+    // CalChart3.1 is more complicated a format so we parse the blocks inline.
+    let offset = 8;
+    if (readFourCharCode(buffer, offset) !== 'SHOW') {
+      throw 'Not a valid CalChart3.3.5 show';
```

Comment:
Should we be more specific with our error messages? This way if somebody imports a show and sends us an email asking why it doesn't work, we'll have a better idea of how to help them.